pkg/sources/git/git.go

```diff
@@ -312,63 +312,80 @@ func GitURLParse(gitURL string) (*url.URL, error) {
-	cloneURL, err := GitURLParse(gitUrl)
-	if err != nil {
-		return "", nil, err
-	}
-	if cloneURL.User == nil {
-		cloneURL.User = userInfo
-	}
 
-	gitArgs := []string{"clone", cloneURL.String(), clonePath}
-	gitArgs = append(gitArgs, args...)
-	cloneCmd := exec.Command("git", gitArgs...)
+	var repo *git.Repository
 
-	safeUrl, err := stripPassword(gitUrl)
-	if err != nil {
-		ctx.Logger().V(1).Info("error stripping password from git url", "error", err)
-	}
-	logger := ctx.Logger().WithValues(
-		"subcommand", "git clone",
-		"repo", safeUrl,
-		"path", clonePath,
-		"args", args,
-	)
+	err = func() error {
+		cloneURL, err := GitURLParse(gitUrl)
+		if err != nil {
+			return err
+		}
+		if cloneURL.User == nil {
+			cloneURL.User = userInfo
+		}
 
-	// Execute command and wait for the stdout / stderr.
-	output, err := cloneCmd.CombinedOutput()
-	if err != nil {
-		err = errors.WrapPrefix(err, "error running 'git clone'", 0)
-	}
-	logger.V(3).Info("git subcommand finished", "output", string(output))
+		gitArgs := []string{"clone", cloneURL.String(), clonePath}
+		gitArgs = append(gitArgs, args...)
+		cloneCmd := exec.Command("git", gitArgs...)
 
-	if cloneCmd.ProcessState == nil {
-		return "", nil, errors.New("clone command exited with no output")
-	}
-	if cloneCmd.ProcessState != nil && cloneCmd.ProcessState.ExitCode() != 0 {
-		logger.V(1).Info("git clone failed", "output", string(output), "error", err)
-		return "", nil, fmt.Errorf("could not clone repo: %s, %w", safeUrl, err)
-	}
+		safeURL, err := stripPassword(gitUrl)
+		if err != nil {
+			ctx.Logger().V(1).Info("error stripping password from git url", "error", err)
+		}
+		logger := ctx.Logger().WithValues(
+			"subcommand", "git clone",
+			"repo", safeURL,
+			"path", clonePath,
+			"args", args,
+		)
 
-	options := &git.PlainOpenOptions{
-		DetectDotGit:          true,
-		EnableDotGitCommonDir: true,
-	}
-	repo, err := git.PlainOpenWithOptions(clonePath, options)
+		// Execute command and wait for the stdout / stderr.
+		output, err := cloneCmd.CombinedOutput()
+		if err != nil {
+			err = errors.WrapPrefix(err, "error running 'git clone'", 0)
+		}
+		logger.V(3).Info("git subcommand finished", "output", string(output))
+
+		if cloneCmd.ProcessState == nil {
+			return fmt.Errorf("clone command exited with no output")
+		}
+		if cloneCmd.ProcessState != nil && cloneCmd.ProcessState.ExitCode() != 0 {
+			logger.V(1).Info("git clone failed", "output", string(output), "error", err)
+			return fmt.Errorf("could not clone repo: %s, %w", safeURL, err)
+		}
+
+		options := &git.PlainOpenOptions{
+			DetectDotGit:          true,
+			EnableDotGitCommonDir: true,
+		}
+		repo, err = git.PlainOpenWithOptions(clonePath, options)
+		if err != nil {
+			return fmt.Errorf("could not open cloned repo: %w", err)
+		}
+		logger.V(1).Info("successfully cloned repo")
+
+		return nil
+	}()
 	if err != nil {
-		return "", nil, fmt.Errorf("could not open cloned repo: %w", err)
+		return "", nil, err
 	}
```

Comment:
Instead of using a `defer`, we could call `CleanOnError` here. The only considerable difference is it wouldn't run if there was a `panic` in the closure.